File: typescript/packages/core/test/unit/http/backgroundInit.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import {
  attachBackgroundInitHandler,
  isFatalStartupInitError,
} from "../../../src/http/backgroundInit";
import { RouteConfigurationError } from "../../../src/http/x402HTTPResourceServer";
import { FacilitatorCapabilityError } from "../../../src/types/facilitator";

describe("backgroundInit", () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe("isFatalStartupInitError", () => {
    it("treats capability and route configuration errors as fatal", () => {
      expect(
        isFatalStartupInitError(new FacilitatorCapabilityError(["upto on solana:devnet: missing"])),
      ).toBe(true);
      expect(
        isFatalStartupInitError(
          new RouteConfigurationError([
            {
              routePattern: "GET /api/generate",
              scheme: "upto",
              network: "solana:devnet",
              reason: "missing_facilitator",
              message: "missing facilitator",
            },
          ]),
        ),
      ).toBe(true);
    });

    it("treats facilitator timeouts as retryable", () => {
      expect(isFatalStartupInitError(new Error("facilitator request timed out"))).toBe(false);
    });
  });

  describe("attachBackgroundInitHandler", () => {
    it("exits the process on a capability mismatch", async () => {
      const exit = vi.spyOn(process, "exit").mockImplementation((() => undefined) as never);
      vi.spyOn(console, "error").mockImplementation(() => undefined);

      attachBackgroundInitHandler(
        Promise.reject(new FacilitatorCapabilityError(["upto on solana:devnet: missing"])),
      );
      await Promise.resolve();
      await Promise.resolve();

      expect(exit).toHaveBeenCalledWith(1);
    });

    it("does not exit on a retryable facilitator timeout", async () => {
      const exit = vi.spyOn(process, "exit").mockImplementation((() => undefined) as never);

      attachBackgroundInitHandler(Promise.reject(new Error("facilitator request timed out")));
      await Promise.resolve();
      await Promise.resolve();

      expect(exit).not.toHaveBeenCalled();
    });
  });
});

File: go/http/background_init_test.go
package http

import (
	"errors"
	"testing"

	x402 "github.com/x402-foundation/x402/go/v2"
)

func TestIsFatalStartupInitErrorTreatsCapabilityAndRouteErrorsAsFatal(t *testing.T) {
	if !IsFatalStartupInitError(x402.NewFacilitatorCapabilityError([]string{"upto on solana:devnet: missing"})) {
		t.Fatal("expected FacilitatorCapabilityError to be fatal")
	}
	if !IsFatalStartupInitError(&RouteConfigurationError{
		Errors: []RouteValidationError{
			{
				RoutePattern: "GET /api/generate",
				Scheme:       "upto",
				Network:      "solana:devnet",
				Reason:       "missing_facilitator",
				Message:      "missing facilitator",
			},
		},
	}) {
		t.Fatal("expected RouteConfigurationError to be fatal")
	}
}

func TestIsFatalStartupInitErrorTreatsTimeoutsAsRetryable(t *testing.T) {
	if IsFatalStartupInitError(errors.New("facilitator request timed out")) {
		t.Fatal("expected facilitator timeouts to be retryable")
	}
}

func TestHandleBackgroundInitErrorExitsOnCapabilityMismatch(t *testing.T) {
	var exitCode int
	called := false
	orig := processExit
	processExit = func(code int) {
		called = true
		exitCode = code
	}
	t.Cleanup(func() { processExit = orig })

	HandleBackgroundInitError(x402.NewFacilitatorCapabilityError([]string{"upto on solana:devnet: missing"}))

	if !called {
		t.Fatal("expected process exit on capability mismatch")
	}
	if exitCode != 1 {
		t.Fatalf("expected exit code 1, got %d", exitCode)
	}
}

func TestHandleBackgroundInitErrorDoesNotExitOnRetryableTimeout(t *testing.T) {
	called := false
	orig := processExit
	processExit = func(int) { called = true }
	t.Cleanup(func() { processExit = orig })

	HandleBackgroundInitError(errors.New("facilitator request timed out"))

	if called {
		t.Fatal("expected no process exit on retryable facilitator timeout")
	}
}
